Godeps/_workspace/src/github.com/docker/distribution/registry/storage/layerwriter.go

```diff
@@ -79,8 +79,26 @@ func (lw *layerWriter) Finish(dgst digest.Digest) (distribution.Layer, error) {
 
 	}
 
-	if err := lw.moveLayer(canonical); err != nil {
-		// TODO(stevvooe): Cleanup?
+	// On NFS filesystem file attributes are cached on client. File may
+	// not exist (os.Stat failes with NotExist) few seconds after the write
+	// has been completed.
+	for retries := 0; ; retries++ {
+		err := lw.moveLayer(canonical)
+		if err == nil {
+			if retries > 0 {
+				ctxu.GetLogger(lw.layerStore.repository.ctx).Debugf("(*layerWriter).Finish: layer moved after %d failed attempts", retries)
+			}
+			break
+		}
+		switch err.(type) {
+		case storagedriver.PathNotFoundError:
```

Comment:
The check for a `source` is performed before a *move* itself and returns `PathNotFoundError` is it doesn't exist. The method doesn't permorm any post move verification.